File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem2_ProductOfArray.scala
package dcp.problems._1_20

/**
 * This problem was asked by Uber.
 *
 * Given an array of integers, return a new array such that each element at index i of the new array is the product of
 * all the numbers in the original array except the one at i.
 *
 * For example, if our input was [1, 2, 3, 4, 5], the expected output would be [120, 60, 40, 30, 24].
 * If our input was [3, 2, 1], the expected output would be [2, 3, 6].
 */

object Problem2_ProductOfArray extends App {

  // brute force O(n^2)
  def solution(arr: Array[Int]): Array[Int] = {
    val result = new Array[Int](arr.length)
    for (i <- arr.indices) {
      var product = 1
      for (j <- arr.indices) {
        if (i != j) {
          product *= arr(j)
        }
      }
      result(i) = product
    }
    result
  }

  // O(n)
  def solution2(arr: Array[Int]): Array[Int] = {
    val cumulativeProduct = arr.foldLeft((List.empty[Int], 1)) { case ((xs, product), elem) =>
      val prod = product * elem
      (xs :+ prod, prod)
    }._1

    val cumulativeProductReverse = arr.reverse.foldLeft((List.empty[Int], 1)) { case ((xs, product), elem) =>
      val prod = product * elem
      (xs :+ prod, prod)
    }._1.reverse

    val result = new Array[Int](arr.length)
    for (i <- arr.indices) {
      if (i == 0) result(i) = cumulativeProductReverse(i + 1)
      else if (i == arr.length - 1) result(i) = cumulativeProduct(i - 1)
      else result(i) = cumulativeProduct(i - 1) * cumulativeProductReverse(i + 1)
    }
    result
  }

}

File: daily-coding-problem/dcp-py/1_20/problem2_productofarray.py
"""
This problem was asked by Uber.

Given an array of integers, return a new array such that each element at index i of the new array is the product of all
the numbers in the original array except the one at i.

For example, if our input was [1, 2, 3, 4, 5], the expected output would be [120, 60, 40, 30, 24]. If our input was
[3, 2, 1], the expected output would be [2, 3, 6].

Follow-up: what if you can't use division?
"""


def product_of_array(ints):
    N = len(ints)

    cumulative_product = 1
    product_prefix = [1] * N
    for i, elem in enumerate(ints):
        cumulative_product *= elem
        product_prefix[i] = cumulative_product

    cumulative_product = 1
    product_prefix_reverse = [1] * N
    for i, elem in enumerate(ints[::-1]):
        cumulative_product *= elem
        product_prefix_reverse[i] = cumulative_product
    product_prefix_reverse = product_prefix_reverse[::-1]

    result = [0] * len(ints)
    for i in range(len(ints)):
        if i == 0:
            result[i] = product_prefix_reverse[i + 1]
        elif i == len(ints) - 1:
            result[i] = product_prefix[i - 1]
        else:
            result[i] = product_prefix[i - 1] * product_prefix_reverse[i + 1]

    return result


def product_of_array_2(ints):
    result = [0] * len(ints)
    for i in range(len(ints)):
        product = 1
        for j in range(len(ints)):
            product *= ints[j] if i != j else 1
        result[i] = product
    return result


assert (product_of_array([1, 2, 3, 4, 5]) == [120, 60, 40, 30, 24])
assert (product_of_array([3, 2, 1]) == [2, 3, 6])
assert (product_of_array_2([1, 2, 3, 4, 5]) == [120, 60, 40, 30, 24])
assert (product_of_array_2([3, 2, 1]) == [2, 3, 6])
